### PR TITLE
github-action: provenance generation

### PR DESCRIPTION
**Pull request by v1v**
## What does this pull request do?

Use https://github.com/github-early-access/generate-build-provenance to create build attestations

### Test

Using a test feature branch and skip upload artifacts or slack notifications:
- release -> https://github.com/elastic/apm-agent-dotnet/actions/runs/9399719545
- release-main -> https://github.com/elastic/apm-agent-dotnet/actions/runs/9399719549/job/25887752695

```diff
diff --git a/.github/workflows/release-main.yml b/.github/workflows/release-main.yml
index d0ef4f1a..b8ec9e29 100644
--- a/.github/workflows/release-main.yml
+++ b/.github/workflows/release-main.yml
@@ -2,7 +2,7 @@ name: release-main
 
 on:
   push:
-    branches: [ "main" ]
+    branches: [ "main", "test/provenance" ]
 
 permissions:
   attestations: write
@@ -36,12 +36,14 @@ jobs:
 
     # Github packages requires authentication, this is likely going away in the future so for now we publish to feedz.io
     - name: publish canary packages to feedz.io
+      if: startsWith(github.ref, 'refs/tags')
       run: dotnet nuget push 'build/output/_packages/*.nupkg' -k ${{ secrets.FEEDZ_IO_API_KEY }} -s ${{ secrets.FEEDZ_IO_API_URL }} --skip-duplicate --no-symbols
       
     - name: publish canary packages github package repository
+      if: startsWith(github.ref, 'refs/tags')
       run: dotnet nuget push 'build/output/_packages/*.nupkg' -k ${{secrets.GITHUB_TOKEN}} -s https://nuget.pkg.github.com/elastic/index.json --skip-duplicate --no-symbols
     
-    - if: ${{ failure() }}
+    - if: ${{ failure() && startsWith(github.ref, 'refs/tags') }}
       uses: elastic/oblt-actions/slack/send@v1.7.0
       with:
         bot-token: ${{ secrets.SLACK_BOT_TOKEN }}
diff --git a/.github/workflows/release.yml b/.github/workflows/release.yml
index 4933a4c4..576d9215 100644
--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -3,6 +3,8 @@ name: release
 on:
   release:
     types: [published]
+  push:
+    branches: [ "test/provenance" ]
 
 permissions:
   contents: write
@@ -39,6 +41,7 @@ jobs:
       run: ./build.sh pack
 
     - name: Release to Nuget
+      if: startsWith(github.ref, 'refs/tags')
       run: .ci/linux/deploy.sh ${{ secrets.NUGET_API_KEY }} ${{ secrets.NUGET_API_URL }}
 
     - name: Set up Docker Buildx
@@ -97,13 +100,14 @@ jobs:
         subject-path: "${{ github.workspace }}/${{ env.PREFIX_APM_PROFILER }}${{ steps.bootstrap.outputs.agent-version }}${{ env.SUFFIX_APM_PROFILER }}"
 
     - name: Attach Profiler And Startup Hooks
+      if: startsWith(github.ref, 'refs/tags')
       continue-on-error: true #continue for now until we see it working in action
       env:
         GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
       run: |
         gh release upload ${{ github.ref_name }} "${{ env.PREFIX_APM_AGENT }}${{ steps.bootstrap.outputs.agent-version }}${{ env.SUFFIX_APM_AGENT }}" "${{ env.PREFIX_APM_PROFILER }}${{ steps.bootstrap.outputs.agent-version }}${{ env.SUFFIX_APM_PROFILER }}"
 
-    - if: ${{ success() }}
+    - if: ${{ success() && startsWith(github.ref, 'refs/tags') }}
       uses: elastic/oblt-actions/slack/send@v1.7.0
       with:
         bot-token: ${{ secrets.SLACK_BOT_TOKEN }}
@@ -113,7 +117,7 @@ jobs:
           Build: (<${{ env.JOB_URL }}|here>)
           Release URL: (<https://github.com/elastic/apm-agent-dotnet/releases/tag/${{ github.ref_name }}|${{ github.ref_name }}>)
 
-    - if: ${{ failure() }}
+    - if: ${{ failure() && startsWith(github.ref, 'refs/tags') }}
       uses: elastic/oblt-actions/slack/send@v1.7.0
       with:
         bot-token: ${{ secrets.SLACK_BOT_TOKEN }}
@@ -152,6 +156,7 @@ jobs:
           subject-path: "${{ github.workspace }}/${{ env.PREFIX_ZIP_FILE }}${{ steps.bootstrap.outputs.agent-version }}${{ env.SUFFIX_ZIP_FILE }}"
 
       - name: Attach Profiler
+        if: startsWith(github.ref, 'refs/tags')
         env:
           GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
         continue-on-error: true #continue for now until we see it working in action
@@ -161,6 +166,7 @@ jobs:
   post-release:
     needs: [ 'release-windows']
     runs-on: ubuntu-latest
+    if: startsWith(github.ref, 'refs/tags')
     env:
       GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
       GIT_TAG: v${{ needs.release-windows.outputs.agent-version }}

```

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -51,22 +58,41 @@ jobs:
...
+      uses: github-early-access/generate-build-provenance@main
+      with:
+        subject-name: "docker.elastic.co/observability/apm-agent-dotnet:${{ steps.bootstrap.outputs.agent-version }}"
+        subject-digest: ${{ steps.push.outputs.digest }}
+        push-to-registry: false
+
+    - name: generate build provenance (APM Agent)
+      uses: github-early-access/generate-build-provenance@main
+      with:
+          subject-path: "${{ github.workspace }}/${{ env.PREFIX_APM_AGENT }}${{ steps.bootstrap.outputs.agent-version }}${{ env.SUFFIX_APM_AGENT }}"
```
Comment (v1v) [inline .github/workflows/release.yml:83]:
```suggestion
        subject-path: "${{ github.workspace }}/${{ env.PREFIX_APM_AGENT }}${{ steps.bootstrap.outputs.agent-version }}${{ env.SUFFIX_APM_AGENT }}"
```

File: .github/workflows/release-main.yml [hunk 2]
```diff
@@ -2,12 +2,13 @@ name: release-main
 
 on:
   push:
-    branches: [ "main" ]
+    branches: [ "main", "feature/provenance" ]
```
Comment (v1v) [inline .github/workflows/release-main.yml:5]:
```suggestion
    branches: [ "main" ]
```

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -51,22 +58,41 @@ jobs:
...
-            --build-arg AGENT_ZIP_FILE=build/output/elastic_apm_profiler_${{ steps.bootstrap.outputs.agent-version }}-linux-x64.zip
-    - name: Push Profiler Docker Image
+    - name: Build and Push Profiler Docker Image
+      id: push
       continue-on-error: true #continue for now until we see it working in action
-      run: |
-        for i in $(seq 1 3); do [ $i -gt 1 ] && sleep 15; docker push docker.elastic.co/observability/apm-agent-dotnet:${{ steps.bootstrap.outputs.agent-version }} && s=0 && break || s=$?; done; (exit $s)
-        
+      uses: docker/build-push-action@v5.3.0
+      with:
```
Comment (reakaleek) [inline .github/workflows/release.yml:65]:
```suggestion
      with:
        cache-from: type=gha
        cache-to: type=gha,mode=max
```

Unrelated optimization I found recently

Comment (reakaleek) [reply .github/workflows/release.yml:65]: Source: https://docs.docker.com/build/cache/backends/gha/#using-dockerbuild-push-action

**Review by reakaleek — APPROVED** [review]

File: .github/workflows/release-main.yml [hunk 4]
```diff
@@ -5,9 +5,10 @@ on:
     branches: [ "main" ]
 
 permissions:
-    contents: write
-    issues: write
-    packages: write
+  contents: write
```
Comment (v1v) [inline .github/workflows/release-main.yml:8]:
```suggestion
  attestations: write
  contents: write
```

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -17,14 +17,21 @@ env:
...
   release:
     runs-on: ubuntu-latest
-
+    env:
+      PREFIX_APM_AGENT: "build/output/ElasticApmAgent_"
+      PREFIX_APM_PROFILER: "build/output/elastic_apm_profiler_"
+      SUFFIX_APM_AGENT: ".zip"
+      SUFFIX_APM_PROFILER: "-linux-x64.zip"
+    permissions:
+      id-token: write
```
Comment (v1v) [inline .github/workflows/release.yml:26]:
```suggestion
      attestations: write
      id-token: write
```

File: .github/workflows/release.yml (hunk 3 above)
Comment (v1v) [reply .github/workflows/release.yml:65]: ah!! I somehow it was some kind of GitHub specific registry but it's not it's about caching in GitHub actions. Thanks :)

**Comment by v1v** [comment]
run docs-build

**Review by reakaleek — DISMISSED** [review]

**Review by Mpdreamz — DISMISSED** [review]

**Review by Mpdreamz — DISMISSED** [review]

**Comment by v1v** [comment]
I'll test this in isolation to validate things work as expected after the merge I ran recently

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -51,23 +60,43 @@ jobs:
...
+
+    - name: Attest image
+      uses: github-early-access/generate-build-provenance@main
+      with:
+        subject-name: ${{ env.DOCKER_IMAGE }}
+        subject-digest: ${{ steps.docker-push.outputs.digest }}
+        push-to-registry: true
+
+    - name: generate build provenance (APM Agent)
+      uses: github-early-access/generate-build-provenance@main
```
Comment (v1v) [inline .github/workflows/release.yml:85]:
```suggestion
      uses: actions/attest-build-provenance@173725a1209d09b31f9d30a3890cf2757ebbff0d  # v1.1.2
```

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -51,23 +60,43 @@ jobs:
...
+      with:
+        subject-path: "${{ github.workspace }}/${{ env.PREFIX_APM_AGENT }}${{ steps.bootstrap.outputs.agent-version }}${{ env.SUFFIX_APM_AGENT }}"
+
+    - name: generate build provenance (APM Profiler)
+      uses: github-early-access/generate-build-provenance@main
```
Comment (v1v) [inline .github/workflows/release.yml:90]:
```suggestion
      uses: actions/attest-build-provenance@173725a1209d09b31f9d30a3890cf2757ebbff0d  # v1.1.2
```

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -110,12 +145,17 @@ jobs:
       - name: Build profiler
         run: ./build.bat profiler-zip
 
+      - name: generate build provenance (APM Profiler)
+        uses: github-early-access/generate-build-provenance@main
```
Comment (v1v) [inline .github/workflows/release.yml:149]:
```suggestion
        uses: actions/attest-build-provenance@173725a1209d09b31f9d30a3890cf2757ebbff0d  # v1.1.2
```

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -39,23 +48,43 @@ jobs:
...
+        cache-to: type=gha,mode=max
+        context: .
+        push: true
+        tags: |
+          ${{ env.DOCKER_IMAGE }}:${{ steps.bootstrap.outputs.agent-version }}
+        build-args: |
+          AGENT_ZIP_FILE=${{ env.PREFIX_APM_PROFILER }}${{ steps.bootstrap.outputs.agent-version }}${{ env.SUFFIX_APM_PROFILER }}
+
+    - name: Attest image
+      uses: github-early-access/generate-build-provenance@main
```
Comment (v1v) [inline .github/workflows/release.yml:66]:
```suggestion
      uses: actions/attest-build-provenance@49df96e17e918a15956db358890b08e61c704919  # v1.2.0
```

**Comment by v1v** [comment]
All good now, it took me a while to verify things worked nice/smoothly... so I feel we are now in a good position to merge this PR.

**Review by Mpdreamz — APPROVED** [review]
Lgtm merge at will :)

**Review by reakaleek — APPROVED** [review]

**Pull request merged**